pkg/algorithm/cell_allocation.go

```diff
@@ -78,6 +79,77 @@ func buddyAlloc(
 	return false
 }
 
+// after buddyAlloc cannot find a healthy cell (or not in suggested nodes),
+// try to split a higher level cell safely to get current level cells
+func safeRelaxedBuddyAlloc(
+	cell *cellBindingPathVertex,
+	freeList ChainCellList,
+	freeCellNum map[CellLevel]int32,
+	currentLevel CellLevel,
+	suggestedNodes common.Set,
+	ignoreSuggestedNodes bool,
+	bindings map[api.CellAddress]*PhysicalCell) bool {
+
+	var splittableCell Cell
+	splittableNum := map[CellLevel]int32{}
+	for i := CellLevel(len(freeList)); i >= CellLevel(1); i-- {
+		// calculate splittable number
+		splittableNum[i] = int32(len(freeList[i])) - freeCellNum[i]
+		if i < CellLevel(len(freeList)) && splittableCell != nil {
+			splittableNum[i] += splittableNum[i+1] * int32(len(splittableCell.GetChildren()))
+		}
+		// iterate higher level cell
+		if splittableCell == nil && len(freeList[i]) > 0 {
+			splittableCell = freeList[i][0]
+		} else if splittableCell != nil {
+			splittableCell = splittableCell.GetChildren()[0]
+		}
+		// check safety
+		if splittableNum[i] < 0 {
+			panic(fmt.Sprintf("VC Safety Broken: level %v cell with free list %v is unsplittable, splittableNum=%v",
+				i, freeList[i], splittableNum[i]))
+		}
+	}
+
+	for l := currentLevel + 1; l <= CellLevel(len(freeList)); l++ {
+		cellNum := int32(len(freeList[l]))
+		if cellNum > splittableNum[l] {
+			cellNum = splittableNum[l]
+		}
+		if cellNum > 0 {
+			splitList := CellList{}
```

Comment:
>splitList [](start = 3, length = 9)

splittableList?

Reply:
the list is used to really split, not splittable only...